pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,334 @@
+                mesh_n[f'{dpf}/fields/{field}/volume_dependent'] = 'false'
+                mesh_n[f'{dpf}/fields/{field}/topology'] = 'mesh'
+
+        self.mesh_n = mesh_n
+
+        # Inital Ascent setup
+        self.a = Ascent()
+        self.ascent_config = ConduitNode(self.conduit)
+        self.ascent_config['mpi_comm'] = comm.py2f()
+        self.ascent_config['runtime/type'] = 'ascent'
+        self.ascent_config['runtine/vtkm/backend'] = 'serial'
+
+        # Open ascent
+        self.a.open(self.ascent_config)
+
+    def __del__(self):
+        if self.a:
+            self.a.close()
+
+    def _config_field(self, cfgsect):
+        cons = self.cfg.items_as('constants', float)
+
+        for k in self.cfg.items(cfgsect):
+            if k.startswith('field-'):
+                field = k.removeprefix('field-')
+                configs = self.cfg.getliteral(cfgsect, k)
+                
+                for kc, vc in configs.items():
+                    if kc == 'field_expr':
+                        if field in self._expr:
+                            raise KeyError(f"Field '{field}' already exists")
+                        self._expr[field] = self._get_expr(vc, subs=cons)
+                    elif kc.startswith('field_expr_'):
+                        sub_field = kc.removeprefix('field_expr_')
+                        if sub_field not in ['x', 'y', 'z'][:self.ndims]:
+                            raise KeyError('Invalid vector subfield ID')
+                        if field not in self._expr:
+                            self._expr[field] = {}
+                        
+                        expr = self._get_expr(vc, subs=cons)
+                        self._expr[field][sub_field] = expr
+
+    def _config_scene(self, cfgsect):
+        cons = self.cfg.items_as('constants', float)
+        scene_cfg = {}
+        self.scenes = ConduitNode(self.conduit)
+
+        i = 0 # scene counter
+        for k in self.cfg.items(cfgsect):
+            if k.startswith('scene-'):
+                configs = self.cfg.getliteral(cfgsect, k)
+                field = configs['field']
+                sname = k.removeprefix('scene-')
+
+                scene_cfg[sname] = configs
+
+                for kc, vc in configs.items():
+                    if kc == 'image_name':
+                        pass
+                    elif kc == 'field_expr':
+                        if field in self._expr:
+                            raise KeyError(f"Field '{field}' already exists")
+                        self._expr[field] = self._get_expr(vc, subs=cons)
+                    elif kc.startswith('field_expr_'):
+                        sub_field = kc.removeprefix('field_expr_')
+                        if sub_field not in ['x', 'y', 'z'][:self.ndims]:
+                            raise KeyError('Invalid vector subfield ID')
+                        if field not in self._expr:
+                            self._expr[field] = {}
```

Comment:
defaultdict

Reply:
I think that wouldn't acheive the same result. I only want values for _expr to be dictionaries in the case when field_expr_... is set, ie when setting the expressions for a vector.